app/javascript/crayons/MentionAutocomplete/MentionAutocomplete.jsx

```diff
@@ -0,0 +1,140 @@
+      textAreaRef.current.value = newValue;
+    },
+    [cursorPlacementData, textAreaRef],
+  );
+
+  const handleSelection = useCallback(
+    (selection) => {
+      const { textBefore, textAfter } = cursorPlacementData;
+      const newValueUntilEndOfSearch = `${textBefore}${selection}`;
+      textAreaRef.current.value = `${newValueUntilEndOfSearch}${textAfter}`;
+
+      const nextCursorPosition = newValueUntilEndOfSearch.length;
+      setIsAutocompleteActive(false);
+      textAreaRef.current.focus();
+      textAreaRef.current.setSelectionRange(
+        nextCursorPosition,
+        nextCursorPosition,
+      );
+    },
+    [cursorPlacementData, textAreaRef],
+  );
+
+  useEffect(() => {
+    const keyEventListener = ({ key }) => {
+      if (key === '@') {
+        if (!shouldKeyPressTriggerSearch(textAreaRef.current)) {
+          return;
+        }
+
+        const textAreaX = textAreaRef.current.offsetLeft;
+        const cursorCoords = getCursorXY(
+          textAreaRef.current,
+          textAreaRef.current.selectionStart,
+        );
+
+        const coords = {
+          y: cursorCoords.y,
+          x: isSmallScreen ? textAreaX : cursorCoords.x,
+        };
+
+        setCursorPlacementData({
+          ...coords,
+          textBefore: textAreaRef.current.value.substring(
+            0,
+            textAreaRef.current.selectionStart,
+          ),
+          textAfter: textAreaRef.current.value.substring(
+            textAreaRef.current.selectionStart,
+          ),
+        });
+        setIsAutocompleteActive(true);
+      }
+    };
+
+    const textArea = textAreaRef.current;
+
+    if (textArea) {
+      textArea.addEventListener('keydown', keyEventListener);
+      return () => textArea.removeEventListener('keydown', keyEventListener);
+    }
+  }, [textAreaRef, isSmallScreen]);
+
+  useEffect(() => {
+    const container = document.getElementById('mention-autocomplete-container');
+    if (!container) {
+      return;
+    }
+    if (isAutocompleteActive) {
+      import('./MentionAutocompleteCombobox').then(
+        ({ MentionAutocompleteCombobox }) => {
+          render(
+            <MentionAutocompleteCombobox
+              onSelect={handleSelection}
+              fetchSuggestions={fetchSuggestions}
+              placementCoords={cursorPlacementData}
+              onSearchTermChange={handleSearchTermChange}
+            />,
+            container,
+          );
+        },
+      );
+    } else {
+      render(null, container);
+    }
+  }, [
+    cursorPlacementData,
+    fetchSuggestions,
+    handleSearchTermChange,
+    handleSelection,
+    isAutocompleteActive,
+  ]);
+
+  return <span id="mention-autocomplete-container" />;
+};
+
+const shouldKeyPressTriggerSearch = (textArea) => {
+  const { selectionStart, value: valueBeforeKeystroke } = textArea;
+
+  const previousCharacter = valueBeforeKeystroke.charAt(selectionStart - 1);
+  const validPreviousCharRegex = new RegExp(/[^A-Za-z0-9]/);
```

Comment:
ignore the `@` keypress if it was directly preceded by a letter or number (e.g. in an email address)

(thanks for the catch @ludwiczakpawel 🙂 )

Reply:
One tiny non blocking detail: since `shouldKeyPressTriggerSearch` is triggered at every key event, would it make sense to transform `validPreviousCharRegex` into a constant outside the function so that we don't instantiate the regular expression each time the function is called?